### PR TITLE
Attempt to canonicalize components for test()/exec().

### PR DESCRIPTION
**Pull request by wanderview**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/pull/54.html" title="Last updated on Jul 20, 2021, 7:52 PM UTC (52d736e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/urlpattern/54/bd70f55...52d736e.html" title="Last updated on Jul 20, 2021, 7:52 PM UTC (52d736e)">Diff</a>

### CODE REVIEW

**Comment by wanderview** [comment]
@domenic here is an attempt at canonicalizing the component values passed to `test()` and `exec()`.  I'm trying to use the URL parser infrastructure for this, but not sure if I got it right.  I wish I could run tests or something against the spec.

**Comment by wanderview** [comment]
@domenic Ok.  The parent PR merged and I updated this a bit to be more correct.  Note:

* I purposely don't use state override for protocol since it prevents the scheme from being set in many cases.
* I forcibly set the scheme in port and pathname depending on it already having been canonicalized before getting passed in to those algorithms.

Not sure if there is a good place to note that kind of stuff in the spec text.

File: spec.bs [hunk 1]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 <div algorithm>
   To <dfn>canonicalize protocol</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. Let |strippedValue| be the given |value| with a single trailing U+003A (:) removed, if any.
+  1. If |type| is "pattern" then return |strippedValue|.
+  1. Let |dummyURL| be a new [=url=] record.
```
Comment (domenic) [inline spec.bs:401]:
```suggestion
  1. Let |dummyURL| be a new [=URL record=].
```

here and elsewhere

File: spec.bs [hunk 2]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
+  1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |strippedValue|, followed by U+003A (:), with |dummyURL| as [=basic URL parser/url=].
```
Comment (domenic) [inline spec.bs]:
```suggestion
  1. Let |parseResult| be the result of running the [=basic URL parser=] given |strippedValue| followed by U+003A (:), with |dummyURL| as [=basic URL parser/url=].
```

this was pretty confusing as I thought you were intentionally going through the full parser.

File: spec.bs (hunk 1 above)
Comment (domenic) [reply spec.bs:401]: It seems like the algorithms are OK with an empty URL record... a bit surprising, but it seems to work. You could double-check with whatwg-url if you wanted.

File: spec.bs [hunk 3]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
 
 <div algorithm>
-  To <dfn>canonicalize port</dfn> given a string |value| and a string |type|:
-
-  1. TODO |value| |type|
+  To <dfn>canonicalize port</dfn> given a string |portValue|, a string |protocolValue|, and a string |type|:
+
+  1. If |type| is "pattern" then return |portValue|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. Set |dummyURL|'s [=url/scheme=] to |protocolValue|.
```
Comment (domenic) [inline spec.bs:440]: I think here and line 452 would be good places to insert such notes.

Mixing `<p class="note">` with the markdown lists can be a bit tricky; you have to match the indent levels.

```
1. Set blah
   <p class="note">blah</p>
1. Blah blah
```

File: spec.bs [hunk 4]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
+  1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |value| with |dummyURL| as [=basic URL parser/url=] and [=hostname state=] as [=basic URL parser/state override=].
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/host=].
 </div algorithm>
 
 <div algorithm>
-  To <dfn>canonicalize port</dfn> given a string |value| and a string |type|:
-
-  1. TODO |value| |type|
+  To <dfn>canonicalize port</dfn> given a string |portValue|, a string |protocolValue|, and a string |type|:
```
Comment (domenic) [inline spec.bs:436]:
```suggestion
  To <dfn>canonicalize port</dfn> given a string |portValue|, a string-or-null |protocolValue|, and a string |type|:
```

Comment (domenic) [reply spec.bs:436]: This algorithm's current call site only gives one argument. I think it should explicitly give null for the second.

File: spec.bs [hunk 5]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
-
-  1. TODO |value| |type|
+  To <dfn>canonicalize pathname</dfn> given a string |pathnameValue|, a string |protocolValue|, and a string |type|:
+
+  1. If |type| is "pattern" then return |pathnameValue|.
+  1. Let |dummyURL| be a new [=url=] record.
+  1. If |protocolValue| is null or the empty string, then set |protocolValue| to "https".
+  1. Set |dummyURL|'s [=url/scheme=] to |protocolValue|.
+  1. If |dummyURL| [=is special=], then:
+    1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |pathnameValue| with |dummyURL| as [=basic URL parser/url=] and [=path start state=] as [=basic URL parser/state override=].
```
Comment (domenic) [inline spec.bs:454]: It's a bit confusing to use two variables both named parseResult for different purposes in these cases. I'd name this one differently.

File: spec.bs [hunk 6]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
+  1. Let |parseResult| be the result of [=Basic URL parser|URL parsing=] |portValue| with |dummyURL| as [=basic URL parser/url=] and [=port state=] as [=basic URL parser/state override=].
+  1. If |parseResult| is failure, then throw a {{TypeError}}.
+  1. Return |dummyURL|'s [=url/port=].
 </div algorithm>
 
 <div algorithm>
-  To <dfn>canonicalize pathname</dfn> given a string |value| and a string |type|:
-
-  1. TODO |value| |type|
+  To <dfn>canonicalize pathname</dfn> given a string |pathnameValue|, a string |protocolValue|, and a string |type|:
```
Comment (domenic) [inline spec.bs:447]: Similar call- site mismatch.

File: spec.bs [hunk 7]
```diff
@@ -386,49 +396,91 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 <div algorithm>
   To <dfn>canonicalize protocol</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. Let |strippedValue| be the given |value| with a single trailing U+003A (:) removed, if any.
+  1. If |type| is "pattern" then return |strippedValue|.
```
Comment (domenic) [inline spec.bs:400]:
```suggestion
  1. If |type| is "`pattern`" then return |strippedValue|.
```

here and elsewhere

**Comment by wanderview** [comment]
Sorry, had to push another change for `canonicalizeProtocol()`.  I did some testing with whatwg-url and realized that I needed to provide `://` and a dummy hostname in order for special schemes to be accepted.  Kind of ugly, but I think its the only way to do this.

File: spec.bs (hunk 2 above)
Comment (domenic) [reply spec.bs]: Here is a good place to insert a note about why using state override doesn't work.

File: spec.bs (hunk 4 above)
Comment (wanderview) [reply spec.bs:436]: This was a mistake.  It should be passing a protocol.

**Comment by wanderview** [comment]
@domenic I think I address the feedback.  I also tested each component in whatwg-url.  I fixed another problem I with cannot-be-a-base-URL pathnames where I had to init the property to empty string.

Also, it seems the bikeshed server is having trouble...

File: spec.bs [hunk 8]
```diff
@@ -386,49 +396,95 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
 <div algorithm>
   To <dfn>canonicalize protocol</dfn> given a string |value| and a string |type|:
 
-  1. TODO |value| |type|
+  1. Let |strippedValue| be the given |value| with a single trailing U+003A (:) removed, if any.
+  1. If |type| is "`pattern`" then return |strippedValue|.
+  1. Let |dummyURL| be a new [=url record=].
```
Comment (domenic) [inline spec.bs:401]:
```suggestion
  1. Let |dummyURL| be a new [=URL record=].
```

here and throughout

File: spec.bs [hunk 9]
```diff
@@ -386,49 +396,95 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
+  1. Let |parseResult| be the result of running the [=basic URL parser=] given |strippedValue| followed by "`://dummy.test`", with |dummyURL| as [=basic URL parser/url=].
```
Comment (domenic) [inline spec.bs:402]:
```suggestion
  1. Let |parseResult| be the result of running the [=basic URL parser=] given |strippedValue| followed by "`://dummy.test`", with |dummyURL| as <i>[=basic URL parser/url=]</i>.
```

Also for `[=basic URL parser/state override=]`. See https://infra.spec.whatwg.org/#algorithm-params and some background discussion in https://github.com/whatwg/infra/pull/325#issuecomment-702373931

File: spec.bs [hunk 10]
```diff
@@ -386,49 +396,95 @@ Each {{URLPattern}} object has an associated <dfn for=URLPattern>hash component<
...
+  1. If |type| is "`pattern`" then return |pathnameValue|.
+  1. Let |dummyURL| be a new [=url record=].
+  1. If |protocolValue| is null or the empty string, then set |protocolValue| to "`https`".
+  1. Set |dummyURL|'s [=url/scheme=] to |protocolValue|.
+     <p class="note">Note, we set the [=url record=]'s [=url/scheme=] in order to check if it [=is special=].  If |protocolValue| is missing we purposely default the URL to being special by using "`https`" for the [=url/scheme=].</p>
+  1. Let |parseResult| be null.
+  1. If |dummyURL| [=is special=], then:
+    1. Set |parseResult| to the result of running [=basic URL parser=] given |pathnameValue| with |dummyURL| as [=basic URL parser/url=] and [=path start state=] as [=basic URL parser/state override=].
+  1. Else:
+    1. Set |dummyURL|'s [=url/path=][`0`] to empty string.
```
Comment (domenic) [inline spec.bs:460]:
```suggestion
    1. Set |dummyURL|'s [=url/path=][0] to empty string.
```

for some reason we typically don't code-ify numbers :)

**Review by domenic — APPROVED** [review]
Looks great with only typographical nits

**Pull request merged**